mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java

```diff
@@ -0,0 +1,301 @@
+        this.quotaManager = quotaManager;
+        this.quotaRootResolver = quotaRootResolver;
+    }
+
+    @Override
+    public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, MessageId messageId, MailboxSession mailboxSession) throws MailboxException {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+        Map<MailboxId, UpdatedFlags> updatedFlags = messageIdMapper.setFlags(newState, replace, messageId);
+        for (Map.Entry<MailboxId, UpdatedFlags> entry : updatedFlags.entrySet()) {
+            dispatchFlagsChange(mailboxSession, entry);
+        }
+    }
+
+    private void dispatchFlagsChange(MailboxSession mailboxSession, Map.Entry<MailboxId, UpdatedFlags> entry) throws MailboxException {
+        Mailbox mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxById(entry.getKey());
+        if (!entry.getValue().getNewFlags().equals(entry.getValue().getOldFlags())) {
+            dispatcher.flagsUpdated(mailboxSession,
+                ImmutableList.of(entry.getValue().getUid()),
+                mailbox,
+                ImmutableList.of(entry.getValue()));
+        }
+    }
+
+    @Override
+    public List<MessageResult> getMessages(List<MessageId> messageIds, final MessageResult.FetchGroup minimal, MailboxSession mailboxSession) throws MailboxException {
+        try {
+            MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.createMessageIdMapper(mailboxSession);
+            return FluentIterable.from(messageIdMapper.find(messageIds, MessageMapper.FetchType.getFetchType(minimal)))
+                .transform(createMessageToResult(minimal))
+                .toList();
+        } catch (RuntimeException runtime) {
+            throw unwrap(runtime);
+        }
+    }
+
+    @Override
+    public void delete(MessageId messageId, final List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+        MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
+        MailboxMapper mailboxMapper = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession);
+
+        Iterable<MetadataWithMailbox> metadatasWithMailbox = FluentIterable
+            .from(messageIdMapper.find(ImmutableList.of(messageId), MessageMapper.FetchType.Metadata))
+            .filter(createIsInMailboxesPredicate(mailboxIds))
+            .transform(EXTRACT_METADATA_FUNCTION);
+
+        messageIdMapper.delete(messageId, mailboxIds);
+
+        for (MetadataWithMailbox metadataWithMailbox: metadatasWithMailbox) {
+            Map<MessageUid, MessageMetaData> map = new HashMap<MessageUid, MessageMetaData>();
+            map.put(metadataWithMailbox.messageMetaData.getUid(), metadataWithMailbox.messageMetaData);
+            dispatcher.expunged(mailboxSession, map, mailboxMapper.findMailboxById(metadataWithMailbox.mailboxId));
```

Comment:
I've got no way to know it. If we want to know it, we have to modify the mapper API. Should we ?